controllers/movement.py

```diff
@@ -104,66 +123,122 @@ def generate_trajectory(self) -> Trajectory:
-
-    # will execute if no other states are executing
-    @default_state
-    def manualdrive(self) -> None:
-        if self.debug_trajectory is True:
-            self.generate_trajectory()
-        if self.drive_local:
-            self.chassis.drive_local(*self.inputs)
-        else:
-            self.chassis.drive_field(*self.inputs)
-
-    @state(first=True)
-    def autodrive(self, state_tm: float, initial_call: bool) -> None:
-        if initial_call:
-            # When the robot is following a trajectory, it doesn't need to re-define the
-            # PID controllers nor does it need to generate another trajectory.
-            # Therefore, this section of code only runs when the state is first active.
-            self.x_controller = PIDController(2.5, 0, 0)
-            self.y_controller = PIDController(2.5, 0, 0)
-            self.heading_controller = ProfiledPIDControllerRadians(
-                5, 0, 0, TrapezoidProfileRadians.Constraints(1, 1)
+        if goal != self.goal:
+            self.goal = goal
+            self.goal_rotation = approach_direction
+
+            self.config = TrajectoryConfig(1, 3)
+            self.config.addConstraint(CentripetalAccelerationConstraint(1.5))
+            topRight = Translation2d(self.goal.X() + 0.25, self.goal.Y() + 0.25)
+            bottomLeft = Translation2d(self.goal.X() - 0.25, self.goal.Y() - 0.25)
+            self.config.addConstraint(
+                RectangularRegionConstraint(
+                    bottomLeft, topRight, MaxVelocityConstraint(0.5)
+                )
             )
-            self.heading_controller.enableContinuousInput(0, math.tau)
 
-            self.drive_controller = HolonomicDriveController(
-                self.x_controller, self.y_controller, self.heading_controller
-            )
-            self.trajectory = self.generate_trajectory()
+    def execute_trajectory(self, trajectory: Trajectory, state_tm: float) -> None:
 
-        target_state = self.trajectory.sample(
+        target_state = trajectory.sample(
             state_tm
         )  # Grabbing the target position at the current point in time from the trajectory.
 
         chassis_speed = self.drive_controller.calculate(
             self.chassis.get_pose(),
             target_state,
-            self.goal.rotation(),  # Calculating the speeds required to get to the target position.
+            # Calculating the speeds required to get to the target position.
+            self.goal.rotation(),
         )
         self.chassis.drive_local(
             chassis_speed.vx,
             chassis_speed.vy,
             chassis_speed.omega,
         )
 
+        self.time_remaining = self.auto_trajectory.totalTime() - state_tm
+
+    def is_at_goal(self) -> bool:
+        return (
+            self.goal.translation() - self.chassis.get_pose().translation()
+        ).norm() < self.POSITION_TOLERANCE and abs(
+            self.goal.rotation().radians() - self.chassis.get_rotation().radians()
```

Comment:
This angle maths is incorrect ­— for example if the goal has a rotation of π and the robot has a current heading of (-π + `ANGLE_TOLERANCE/2`), this will say this is _not_ within the tolerance, despite the robot being in less than `ANGLE_TOLERANCE` away from the desired heading.